File: May-LeetCoding-Challenge/03-Ransom-Note/Ransom-Note.py
class Solution1:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        if ransomNote == magazine:
            return True
        else:
            for r in ransomNote:
                if r in magazine:
                    magazine = magazine.replace(r, '', 1)
                else:
                    return False
            return True


class Solution2:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        if ransomNote == magazine:
            return True
        else:
            magazines_dict = {}
            for m in magazine:
                if m in magazines_dict:
                    magazines_dict[m] += 1
                else:
                    magazines_dict[m] = 1

            for r in ransomNote:
                if r in magazines_dict and magazines_dict[r] > 0:
                    magazines_dict[r] -= 1
                else:
                    return False
            return True

File: May-LeetCoding-Challenge/03-Ransom-Note/Ransom-Note.scala
//solution typical "scala" way
object Solution1 {
  def canConstruct(ransomNote: String, magazine: String): Boolean =
    magazine.diff(ransomNote).length == magazine.length - ransomNote.length
}


object Solution2 {
  def canConstruct(ransomNote: String, magazine: String): Boolean = canConstructRec(ransomNote, magazine, rIndex = 0, alwaysPresent = true)

  def canConstructRec(ransomNote: String, magazine: String, rIndex: Int, alwaysPresent: Boolean): Boolean = {
    if (!alwaysPresent) false
    else if (ransomNote.length <= rIndex) true
    else {
      var mIndex = 0
      while (mIndex <= magazine.length - 1 && ransomNote(rIndex) != magazine(mIndex)) mIndex = mIndex + 1
      canConstructRec(ransomNote, removeByIndex(magazine, mIndex), rIndex + 1, !(mIndex == magazine.length))
    }
  }

  //remove character from text by index
  def removeByIndex(list: String, index: Int): String = list.take(index) ++ list.drop(index + 1)

}